File: internal/mods/dependencies.go
package mods

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/Masterminds/semver/v3"
	"junimohut/internal/archive"
)

const (
	DependencyMissing       = "missing"
	DependencyVersionTooLow = "version_too_low"
	DependencyDisabled      = "disabled"
)

type depEntry struct {
	UniqueID       string
	MinimumVersion string
	IsRequired     bool
	IsContentPack  bool
}

// ResolveDependencies checks each mod's manifest dependencies against the installed library.
func ResolveDependencies(mods []Mod) []Mod {
	byUniqueID := modIndexByUniqueID(mods)

	out := make([]Mod, len(mods))
	for i, m := range mods {
		issues := resolveModDependencies(m, byUniqueID)
		m.DependencyIssues = issues
		m.MissingDependencyCount = len(issues)
		out[i] = m
	}
	return out
}

// ResolveManifestDependencies checks a manifest against the installed library.
func ResolveManifestDependencies(manifest Manifest, library []Mod) []DependencyIssue {
	mod := Mod{Manifest: manifest}
	return resolveModDependencies(mod, modIndexByUniqueID(library))
}

func modIndexByUniqueID(mods []Mod) map[string]Mod {
	byUniqueID := make(map[string]Mod, len(mods))
	for _, m := range mods {
		uid := m.Manifest.UniqueID
		if uid == "" {
			continue
		}
		if _, ok := byUniqueID[uid]; !ok {
			byUniqueID[uid] = m
		}
	}
	return byUniqueID
}

func resolveModDependencies(mod Mod, byUniqueID map[string]Mod) []DependencyIssue {
	entries := collectDependencyEntries(mod)
	if len(entries) == 0 {
		return nil
	}

	selfID := mod.Manifest.UniqueID
	issues := make([]DependencyIssue, 0, len(entries))
	for _, entry := range entries {
		if entry.UniqueID == "" || entry.UniqueID == selfID {
			continue
		}

		provider, ok := byUniqueID[entry.UniqueID]
		if !ok {
			if entry.IsRequired {
				issues = append(issues, newDependencyIssue(entry, DependencyMissing, nil))
			}
			continue
		}

		installedVersion := provider.Manifest.Version
		if entry.MinimumVersion != "" && !versionSatisfies(installedVersion, entry.MinimumVersion) {
			if entry.IsRequired {
				issues = append(issues, newDependencyIssue(entry, DependencyVersionTooLow, &provider))
			}
			continue
		}
		if !provider.Enabled && entry.IsRequired {
			issues = append(issues, newDependencyIssue(entry, DependencyDisabled, &provider))
		}
	}
	return issues
}

func newDependencyIssue(entry depEntry, state string, provider *Mod) DependencyIssue {
	issue := DependencyIssue{
		UniqueID:       entry.UniqueID,
		MinimumVersion: entry.MinimumVersion,
		IsRequired:     entry.IsRequired,
		IsContentPack:  entry.IsContentPack,
		State:          state,
	}
	if provider != nil {
		issue.InstalledName = provider.Manifest.Name
		issue.InstalledVersion = provider.Manifest.Version
		issue.ProviderModID = provider.ID
		issue.NexusModID = nexusModIDFromManifest(provider.Manifest)
	}
	return issue
}

func nexusModIDFromManifest(m Manifest) string {
	for _, key := range m.UpdateKeys {
		if strings.HasPrefix(key, "Nexus:") {
			return strings.TrimPrefix(key, "Nexus:")
		}
	}
	return ""
}

func collectDependencyEntries(mod Mod) []depEntry {
	seen := map[string]depEntry{}

	if cp := mod.Manifest.ContentPackFor; cp != nil && cp.UniqueID != "" {
		seen[cp.UniqueID] = depEntry{
			UniqueID:       cp.UniqueID,
			MinimumVersion: cp.MinimumVersion,
			IsRequired:     true,
			IsContentPack:  true,
		}
	}

	for _, dep := range mod.Manifest.Dependencies {
		if dep.UniqueID == "" {
			continue
		}
		required := dep.IsRequired == nil || bool(*dep.IsRequired)
		existing, ok := seen[dep.UniqueID]
		if !ok {
			seen[dep.UniqueID] = depEntry{
				UniqueID:       dep.UniqueID,
				MinimumVersion: dep.MinimumVersion,
				IsRequired:     required,
			}
			continue
		}
		if required && !existing.IsRequired {
			existing.IsRequired = true
		}
		if existing.MinimumVersion == "" && dep.MinimumVersion != "" {
			existing.MinimumVersion = dep.MinimumVersion
		}
		seen[dep.UniqueID] = existing
	}

	out := make([]depEntry, 0, len(seen))
	for _, entry := range seen {
		out = append(out, entry)
	}
	return out
}

func versionSatisfies(installedVersion, minimumVersion string) bool {
	minimumVersion = strings.TrimSpace(minimumVersion)
	if minimumVersion == "" {
		return true
	}
	installedVersion = strings.TrimSpace(installedVersion)
	if installedVersion == "" {
		return false
	}

	installed, err1 := semver.NewVersion(strings.TrimPrefix(installedVersion, "v"))
	minimum, err2 := semver.NewVersion(strings.TrimPrefix(minimumVersion, "v"))
	if err1 != nil || err2 != nil {
		return true
	}
	return !installed.LessThan(minimum)
}

// PreviewInstallDependencies extracts manifests from archives and checks dependencies.
func PreviewInstallDependencies(archivePaths []string, library []Mod) ([]InstallDependencyPreview, error) {
	library = DedupeByUniqueID(DedupeByID(library))
	var previews []InstallDependencyPreview
	for _, archivePath := range archivePaths {
		manifests, err := extractManifestsFromArchive(archivePath)
		if err != nil {
			return nil, fmt.Errorf("%s: %w", filepath.Base(archivePath), err)
		}
		for _, manifest := range manifests {
			issues := ResolveManifestDependencies(manifest, library)
			if len(issues) == 0 {
				continue
			}
			previews = append(previews, InstallDependencyPreview{
				ArchivePath: archivePath,
				ModName:     manifest.Name,
				UniqueID:    manifest.UniqueID,
				Issues:      issues,
			})
		}
	}
	return previews, nil
}

func extractManifestsFromArchive(archivePath string) ([]Manifest, error) {
	tmpDir, err := os.MkdirTemp("", "sdvm-preview-*")
	if err != nil {
		return nil, err
	}
	defer os.RemoveAll(tmpDir)

	if err := archive.Extract(archivePath, tmpDir); err != nil {
		return nil, fmt.Errorf("extract archive: %w", err)
	}

	paths, err := findAllManifests(tmpDir)
	if err != nil {
		return nil, err
	}
	paths = FilterRootManifests(paths, tmpDir)
	if len(paths) == 0 {
		return nil, fmt.Errorf("no manifest.json found in archive")
	}

	manifests := make([]Manifest, 0, len(paths))
	for _, path := range paths {
		manifest, err := ParseManifest(path)
		if err != nil {
			return nil, err
		}
		manifests = append(manifests, manifest)
	}
	return manifests, nil
}

File: frontend/src/lib/mods/dependencies.ts
import type { Mod } from "$lib/api/client";

export type DependencyRow = {
  uniqueID: string;
  minimumVersion: string;
  isRequired: boolean;
  isContentPack: boolean;
  state: "satisfied" | "missing" | "version_too_low" | "disabled" | "optional";
  installedName?: string;
  installedVersion?: string;
  providerModId?: string;
  nexusModId?: string;
};

export type DependentRow = {
  modId: string;
  name: string;
  uniqueID: string;
  version: string;
  enabled: boolean;
  minimumVersion: string;
  isRequired: boolean;
  isContentPack: boolean;
  nexusModId?: string;
};

export type InstallDependencyPreview = {
  archivePath: string;
  modName: string;
  uniqueID: string;
  issues: NonNullable<Mod["dependencyIssues"]>;
};

type DepEntry = {
  uniqueID: string;
  minimumVersion: string;
  isRequired: boolean;
  isContentPack: boolean;
};

function parseVersion(version: string): number[] | null {
  const cleaned = version.trim().replace(/^v/i, "");
  const parts = cleaned.split(/[.-]/).map((p) => parseInt(p, 10));
  if (parts.some((n) => Number.isNaN(n))) return null;
  return parts;
}

function versionSatisfies(
  installedVersion: string,
  minimumVersion: string,
): boolean {
  const minimum = minimumVersion.trim();
  if (!minimum) return true;
  const installed = installedVersion.trim();
  if (!installed) return false;

  const instParts = parseVersion(installed);
  const minParts = parseVersion(minimum);
  if (!instParts || !minParts) return true;

  const len = Math.max(instParts.length, minParts.length);
  for (let i = 0; i < len; i++) {
    const a = instParts[i] ?? 0;
    const b = minParts[i] ?? 0;
    if (a > b) return true;
    if (a < b) return false;
  }
  return true;
}

function nexusModIDFromManifest(mod: Mod): string {
  for (const key of mod.manifest?.UpdateKeys ?? []) {
    if (key.startsWith("Nexus:")) return key.slice("Nexus:".length);
  }
  return "";
}

function canonicalUniqueID(uid: string): string {
  return uid.toLowerCase();
}

function uniqueIDsEqual(a: string, b: string): boolean {
  return a.localeCompare(b, undefined, { sensitivity: "accent" }) === 0;
}

function collectDependencyEntries(mod: Mod): DepEntry[] {
  const seen = new Map<string, DepEntry>();

  const contentPack = mod.manifest?.ContentPackFor;
  if (contentPack?.UniqueID) {
    seen.set(canonicalUniqueID(contentPack.UniqueID), {
      uniqueID: contentPack.UniqueID,
      minimumVersion: contentPack.MinimumVersion ?? "",
      isRequired: true,
      isContentPack: true,
    });
  }

  for (const dep of mod.manifest?.Dependencies ?? []) {
    if (!dep.UniqueID) continue;
    const required = dep.IsRequired ?? true;
    const key = canonicalUniqueID(dep.UniqueID);
    const existing = seen.get(key);
    if (!existing) {
      seen.set(key, {
        uniqueID: dep.UniqueID,
        minimumVersion: dep.MinimumVersion ?? "",
        isRequired: required,
        isContentPack: false,
      });
      continue;
    }
    if (required && !existing.isRequired) existing.isRequired = true;
    if (!existing.minimumVersion && dep.MinimumVersion) {
      existing.minimumVersion = dep.MinimumVersion;
    }
  }

  return [...seen.values()];
}

function newDependencyIssue(
  entry: DepEntry,
  state: NonNullable<Mod["dependencyIssues"]>[number]["state"],
  provider?: Mod,
): NonNullable<Mod["dependencyIssues"]>[number] {
  const issue: NonNullable<Mod["dependencyIssues"]>[number] = {
    uniqueID: entry.uniqueID,
    minimumVersion: entry.minimumVersion,
    isRequired: entry.isRequired,
    isContentPack: entry.isContentPack,
    state,
  };
  if (provider) {
    issue.installedName = provider.manifest?.Name;
    issue.installedVersion = provider.manifest?.Version;
    issue.providerModId = provider.id;
    issue.nexusModId = nexusModIDFromManifest(provider);
  }
  return issue;
}

/** Client-side dependency resolution for mock mode and detail-pane rows. */
export function resolveDependencies(mods: Mod[]): Mod[] {
  const byUniqueID = modIndexByUniqueID(mods);

  return mods.map((mod) => {
    const issues = resolveModIssues(mod, byUniqueID);
    return {
      ...mod,
      dependencyIssues: issues,
      missingDependencyCount: issues.length,
    };
  });
}

function modIndexByUniqueID(mods: Mod[]): Map<string, Mod> {
  const byUniqueID = new Map<string, Mod>();
  for (const mod of mods) {
    registerModInUniqueIDIndex(byUniqueID, mod.manifest?.UniqueID ?? "", mod);
    for (const uid of mod.packSiblingUIDs ?? []) {
      registerModInUniqueIDIndex(byUniqueID, uid, mod);
    }
  }
  return byUniqueID;
}

function registerModInUniqueIDIndex(
  byUniqueID: Map<string, Mod>,
  uid: string,
  mod: Mod,
): void {
  if (!uid) return;
  const key = canonicalUniqueID(uid);
  if (byUniqueID.has(key)) return;
  byUniqueID.set(key, mod);
}

function resolveModIssues(
  mod: Mod,
  byUniqueID: Map<string, Mod>,
): NonNullable<Mod["dependencyIssues"]> {
  const selfID = mod.manifest?.UniqueID ?? "";
  const issues: NonNullable<Mod["dependencyIssues"]> = [];

  for (const entry of collectDependencyEntries(mod)) {
    if (!entry.uniqueID || uniqueIDsEqual(entry.uniqueID, selfID)) continue;

    const provider = byUniqueID.get(canonicalUniqueID(entry.uniqueID));
    if (!provider) {
      if (entry.isRequired) {
        issues.push(newDependencyIssue(entry, "missing"));
      }
      continue;
    }

    const installedVersion = provider.manifest?.Version ?? "";
    if (
      entry.minimumVersion &&
      !versionSatisfies(installedVersion, entry.minimumVersion)
    ) {
      if (entry.isRequired) {
        issues.push(newDependencyIssue(entry, "version_too_low", provider));
      }
      continue;
    }
    if (!provider.enabled && entry.isRequired) {
      issues.push(newDependencyIssue(entry, "disabled", provider));
    }
  }

  return issues;
}

function issueToRowState(state: string): DependencyRow["state"] {
  if (state === "version_too_low") return "version_too_low";
  if (state === "disabled") return "disabled";
  if (state === "missing") return "missing";
  return "missing";
}

export function dependencyRowsForMod(
  mod: Mod,
  allMods: Mod[],
): DependencyRow[] {
  const selfID = mod.manifest?.UniqueID ?? "";
  const byUniqueID = modIndexByUniqueID(allMods);

  const issueByID = new Map(
    (mod.dependencyIssues ?? []).map((issue) => [issue.uniqueID, issue]),
  );

  return collectDependencyEntries(mod).map((entry) => {
    const issue = issueByID.get(entry.uniqueID);
    if (issue) {
      return {
        uniqueID: entry.uniqueID,
        minimumVersion: entry.minimumVersion,
        isRequired: entry.isRequired,
        isContentPack: entry.isContentPack,
        state: issueToRowState(issue.state),
        installedName: issue.installedName,
        installedVersion: issue.installedVersion,
        providerModId: issue.providerModId,
        nexusModId: issue.nexusModId,
      };
    }

    const provider = byUniqueID.get(canonicalUniqueID(entry.uniqueID));
    if (provider && !uniqueIDsEqual(entry.uniqueID, selfID)) {
      return {
        uniqueID: entry.uniqueID,
        minimumVersion: entry.minimumVersion,
        isRequired: entry.isRequired,
        isContentPack: entry.isContentPack,
        state: "satisfied",
        installedName: provider.manifest?.Name,
        installedVersion: provider.manifest?.Version,
        providerModId: provider.id,
        nexusModId: nexusModIDFromManifest(provider),
      };
    }

    return {
      uniqueID: entry.uniqueID,
      minimumVersion: entry.minimumVersion,
      isRequired: entry.isRequired,
      isContentPack: entry.isContentPack,
      state: entry.isRequired ? "missing" : "optional",
    };
  });
}

/** Mods in the library that declare a dependency on the given mod's Unique ID. */
export function dependentRowsForMod(mod: Mod, allMods: Mod[]): DependentRow[] {
  const providerID = mod.manifest?.UniqueID;
  if (!providerID) return [];

  const rows: DependentRow[] = [];
  for (const candidate of allMods) {
    if (candidate.id === mod.id) continue;

    for (const entry of collectDependencyEntries(candidate)) {
      if (!uniqueIDsEqual(entry.uniqueID, providerID)) continue;
      rows.push({
        modId: candidate.id,
        name: candidate.manifest?.Name ?? candidate.folderPath,
        uniqueID: candidate.manifest?.UniqueID ?? "",
        version: candidate.manifest?.Version ?? "",
        enabled: candidate.enabled,
        minimumVersion: entry.minimumVersion,
        isRequired: entry.isRequired,
        isContentPack: entry.isContentPack,
        nexusModId: nexusModIDFromManifest(candidate) || undefined,
      });
      break;
    }
  }

  return rows.sort((a, b) => a.name.localeCompare(b.name));
}

export function countModsWithDependencyIssues(mods: Mod[]): number {
  return mods.filter((m) => (m.missingDependencyCount ?? 0) > 0).length;
}

export function nexusSearchUrl(uniqueID: string): string {
  return `https://www.nexusmods.com/stardewvalley/mods/?keyword=${encodeURIComponent(uniqueID)}`;
}

export function nexusModPageUrl(nexusModId: string): string {
  return `https://www.nexusmods.com/stardewvalley/mods/${nexusModId}`;
}

export function previewInstallDependenciesMock(
  archivePaths: string[],
  library: Mod[],
): InstallDependencyPreview[] {
  void archivePaths;
  return library
    .filter((m) => (m.missingDependencyCount ?? 0) > 0)
    .slice(0, 1)
    .map((m) => ({
      archivePath: archivePaths[0] ?? "",
      modName: m.manifest?.Name ?? m.folderPath,
      uniqueID: m.manifest?.UniqueID ?? "",
      issues: m.dependencyIssues ?? [],
    }));
}
